src/net.h

```diff
@@ -703,6 +705,55 @@ class V1TransportDeserializer final : public TransportDeserializer
     CNetMessage GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time) override;
 };
 
+class V2TransportDeserializer : public TransportDeserializer
+{
+private:
+    std::unique_ptr<ChaCha20Poly1305AEAD> m_aead;
+    uint64_t m_payload_seqnr; // sequence number for the payload
+    uint64_t m_aad_seqnr;     // sequence number for the packet length (AD)
+    int m_aad_pos;            // position in the aad keystream
+    bool m_in_data;           // parsing header (false) or data (true)
+    uint32_t m_message_size;  // expected message size
+    CDataStream vRecv;        // received message data
+    unsigned int m_hdr_pos;   // read pos in header
+    unsigned int m_data_pos;  // read pos in data
+
+    int readHeader(const char* pch, unsigned int nBytes);
+    int readData(const char* pch, unsigned int nBytes);
+
+public:
+    V2TransportDeserializer(const CMessageHeader::MessageStartChars& pchMessageStartIn, const CPrivKey& k1, const CPrivKey& k2) : m_aead(new ChaCha20Poly1305AEAD(k1.data(), k1.size(), k2.data(), k2.size())), m_payload_seqnr(0), m_aad_seqnr(0), m_aad_pos(0), vRecv(SER_NETWORK, INIT_PROTO_VERSION)
```

Comment:
Perhaps I'm mistaken, but it seems like this `pchMessageStartIn` is not needed. I tried compiling with it removed and that seemed to work.